x/data/server/msg_server.go

```diff
@@ -123,3 +126,54 @@ func (s serverImpl) SignData(goCtx context.Context, request *data.MsgSignData) (
 
 	return &data.MsgSignDataResponse{}, nil
 }
+
+func (s serverImpl) DefineResolver(ctx context.Context, msg *data.MsgDefineResolver) (*data.MsgDefineResolverResponse, error) {
+	manager, err := cosmossdk.AccAddressFromBech32(msg.Manager)
+	if err != nil {
+		return nil, err
+	}
+
+	id, err := s.stateStore.ResolverInfoStore().InsertReturningID(ctx, &datav1alpha2.ResolverInfo{
+		Url:     msg.ResolverUrl,
+		Manager: manager.Bytes(),
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return &data.MsgDefineResolverResponse{ResolverId: id}, nil
+}
+
+func (s serverImpl) RegisterResolver(ctx context.Context, msg *data.MsgRegisterResolver) (*data.MsgRegisterResolverResponse, error) {
+	resolverInfo, err := s.stateStore.ResolverInfoStore().Get(ctx, msg.ResolverId)
+	if err != nil {
+		return nil, err
+	}
+
+	manager, err := cosmossdk.AccAddressFromBech32(msg.Manager)
+	if err != nil {
+		return nil, err
+	}
+
+	if !bytes.Equal(resolverInfo.Manager, manager.Bytes()) {
+		return nil, data.ErrUnauthorizedResolverManager
+	}
+
+	for _, datum := range msg.Data {
+		_, id, _, err := s.anchorAndGetIRI(sdk.UnwrapSDKContext(ctx), datum)
+		if err != nil {
+			return nil, err
+		}
+		err = s.stateStore.DataResolverStore().Save(
+			ctx,
+			&datav1alpha2.DataResolver{
+				ResolverId: msg.ResolverId,
+				DataId:     id,
```

Comment:
If you choose to change above..

```suggestion
				Id:     id,
```